File: erp_bot/src/knowledge/domain_router.py
"""
Domain router — classify user intent BEFORE web search or generic LLM.
Mirrors src/lib/ekhata/domainRouter.ts for Python backend parity.
"""

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import Literal

from .nepal_ai_runtime import check_safety, is_accounting_question, is_incomplete_transaction

ConversationMode = Literal[
    "journal_entry",
    "accounting_qa",
    "framework_qa",
    "compliance_qa",
    "meta_system",
    "emotional_chat",
    "external_fact",
    "report",
    "correction",
    "education",
    "casual",
]

ACCOUNTING_TERMS = re.compile(
    r"\b(sampati|sampatti|asset|liability|equity|puni|dayitwo|rin|aamdani|kharcha|"
    r"debit|credit|journal|ledger|voucher|hisab|lekha|khata|udhaar|receivable|payable|"
    r"provision|accrual|depreciation|goodwill|inventory|stock|cogs|vat|tds|ssf|gratuity|"
    r"salary|bikri|becheko|kharid|kineko|tiryo|capital|loan|drawings|trial\s*balance|"
    r"ifrs|nas|nfrs|faithful\s*representation|recognition|measurement|fair\s*value|"
    r"going\s*concern|audit|fiscal|shrawan|ashadh|ird|pan\b|accounting|hisabkitab)\b|[\u0900-\u097F]",
    re.I,
)

FRAMEWORK_TERMS = re.compile(
    r"\b(faithful\s*representation|recognition\s*criteria|conceptual\s*framework|"
    r"qualitative\s*characteristics|derecognition|comprehensive\s*income|"
    r"para\s*\d+\.\d+|sambandhitata|manyata|mulyankan)\b",
    re.I,
)

COMPLIANCE_TERMS = re.compile(
    r"\b(ird|vat\s*act|income\s*tax|tds\s*rate|ssf\s*rate|fiscal\s*year|"
    r"tax\s*invoice|withholding|compliance|statutory)\b",
    re.I,
)

ENTRY_SIGNALS = re.compile(
    r"\b(sold|sale|bought|purchase|paid|received|tiryo|tireko|diye|diyeko|kineko|kineye|kinye|becheko|bikri|"
    r"liyo|liye|liya|gare|garyo|garya|replace|repair|"
    r"kharcha|noksan|nokshan|ghateko|ghatyo|loss|salary|vat|tds|loan|capital|return|commission|advance|jama|"
    r"consume|khaye|nikale|istimal|prayog|invest|lagani|haleko|drawings)\b.*\d|"
    r"\d.*\b(sold|sale|bought|purchase|paid|received|tiryo|tireko|diye|diyeko|kineko|kineye|kinye|becheko|"
    r"liyo|liye|liya|gare|garyo|garya|replace|repair|"
    r"kharcha|noksan|nokshan|ghateko|loss|salary|vat|tds|loan|capital|return|commission|advance|"
    r"consume|khaye|nikale|istimal|prayog|invest|lagani|haleko)\b",
    re.I,
)

OWNER_TRANSACTION = re.compile(
    r"\b(aafai|aafaile|afai|afno|owner|malik|personal|nijee|mero)\b.*\b(consume|khaye|liye|use|istimal|nikale|invest|lagani|haleko)\b|"
    r"\b(consume|khaye|liye|use|istimal|nikale|invest|lagani)\b.*\b(aafai|aafaile|afai|afno|owner|malik|personal)\b|"
    r"\b(pasal|shop|business)\s+(?:ko\s+)?(?:saman|goods|mal)\b.*\d",
    re.I,
)

INCOMPLETE_ENTRY = re.compile(
    r"\b(tiryo|tireko|diyo|diyeko|becheko|bikri|kineko|kineye|kinye|kharid|kharcha|noksan|"
    r"liyo|liye|liya|gare|garyo|garya|replace|repair|"
    r"paid|received|bech|kin|jama|aayo|consume|khaye|nikale)\b",
    re.I,
)

QUESTION = re.compile(
    r"\b(what|how|why|when|where|who|k\s*ho|ke\s*ho|kasari|kina|define|explain|"
    r"meaning|matlab|difference|compare|classify|entry|journal|vana|bana|bujhau|"
    r"sikau|artha|bujha|samjha)\b|\?",
    re.I,
)

FOLLOW_UP = re.compile(
    r"\b(vana|bana|bujhau|samjha|explain|accounting\s*ma|properly|clear|detail|"
    r"thik\s*le|ramrari|feri)\b",
    re.I,
)

REPORT = re.compile(
    r"\b(report|summary|statement|ledger|day\s*book|trial\s*balance|p\s*&\s*l|profit)\b",
    re.I,
)

EDUCATION = re.compile(
    r"\b(explain|sikau|bujhau|what\s+is|ke\s+ho|meaning|artha|kasari|teach|sikha)\b",
    re.I,
)

CORRECTION = re.compile(
    r"\b(wrong|galat|fix|sudhar|cancel|hatau|last\s+entry|reverse|ulto)\b",
    re.I,
)

EXTERNAL_FACT = re.compile(
    r"\b(weather|mausam|news|population|prime\s*minister|football|cricket|movie)\b",
    re.I,
)


@dataclass
class DomainRoute:
    mode: ConversationMode
    confidence: float
    block_web_search: bool


def classify_domain(text: str) -> DomainRoute:
    """Classify message domain with web-search guard."""
    t = (text or "").strip()
    if not t:
        return DomainRoute("casual", 1.0, True)

    if check_safety(t):
        return DomainRoute("casual", 0.95, True)

    if is_accounting_question(t):
        if COMPLIANCE_TERMS.search(t) or re.search(r"\b(vat|tds|ssf|tax|kar)\b", t, re.I):
            return DomainRoute("compliance_qa", 0.9, True)
        return DomainRoute("accounting_qa", 0.92, True)

    if CORRECTION.search(t):
        return DomainRoute("correction", 0.9, True)
    if REPORT.search(t):
        return DomainRoute("report", 0.88, True)

    if ENTRY_SIGNALS.search(t) and not QUESTION.search(t):
        return DomainRoute("journal_entry", 0.9, True)

    if OWNER_TRANSACTION.search(t) and re.search(r"\d", t):
        return DomainRoute("journal_entry", 0.92, True)

    if INCOMPLETE_ENTRY.search(t) and not re.search(r"\d", t) and not QUESTION.search(t):
        return DomainRoute("journal_entry", 0.88, True)

    if is_incomplete_transaction(t):
        return DomainRoute("journal_entry", 0.86, True)

    if FRAMEWORK_TERMS.search(t):
        return DomainRoute("framework_qa", 0.88, True)
    if COMPLIANCE_TERMS.search(t):
        return DomainRoute("compliance_qa", 0.85, True)

    if ACCOUNTING_TERMS.search(t):
        if QUESTION.search(t) or FOLLOW_UP.search(t):
            return DomainRoute("accounting_qa", 0.88, True)
        if re.search(r"\d", t):
            return DomainRoute("journal_entry", 0.75, True)
        return DomainRoute("accounting_qa", 0.7, True)

    if FOLLOW_UP.search(t):
        return DomainRoute("accounting_qa", 0.8, True)

    if EDUCATION.search(t) and not ENTRY_SIGNALS.search(t):
        return DomainRoute("education", 0.85, True)

    if EXTERNAL_FACT.search(t):
        return DomainRoute("external_fact", 0.75, False)

    if QUESTION.search(t):
        return DomainRoute("accounting_qa", 0.55, True)

    return DomainRoute("casual", 0.4, True)


def should_block_web_search(text: str) -> bool:
    return classify_domain(text).block_web_search

File: src/lib/ekhata/domainRouter.ts
/**
 * e-Khata Domain Router — classifies user intent domain BEFORE web search or LLM.
 * Prevents accounting terms (sampatti, capital, provision) from hitting Wikipedia.
 */

import { isSemanticTransaction, parseSemanticFrame } from "./semanticNepaliBrain";
import { isNepaliAccountingQuestion } from "../nepal-ai/questionDetect";

export type EKhataDomain =
  | "journal_entry"
  | "accounting_qa"
  | "framework_qa"
  | "compliance_qa"
  | "meta_system"
  | "emotional_chat"
  | "external_fact";

export interface DomainRouteResult {
  domain: EKhataDomain;
  confidence: number;
  /** Never search Wikipedia when true */
  blockWebSearch: boolean;
}

const ACCOUNTING_TERMS =
  /\b(sampati|sampatti|asset|assets|liability|liabilities|equity|puni|dayitwo|rin|aamdani|kharcha|kharcho|debit|credit|dr\b|cr\b|journal|ledger|voucher|hisab|lekha|khata|udhaar|udhar|udharo|receivable|payable|debtor|creditor|provision|accrual|outstanding|prepaid|depreciation|amortization|impairment|goodwill|inventory|stock|cogs|vat|tds|ssf|gratuity|talab|salary|payroll|bikri|becheko|kharid|kineko|tiryo|jama|drawings|capital|loan|contra|reconciliation|trial\s*balance|balance\s*sheet|profit|loss|turnover|working\s*capital|cash\s*flow|ifrs|nas|nfrs|gaap|double\s*entry|write\s*off|bad\s*debt|commission|bhaada|bhada|rent|interest|dividend|opening\s*balance|sales\s*return|purchase\s*return|credit\s*note|debit\s*note|advance|prepaid|fair\s*value|going\s*concern|faithful\s*representation|recognition|measurement|consolidation|audit|revaluation|impairment|deferred|contingent|amortization|ebitda|ebit|gross|net\s*profit|retained\s*earnings|petty\s*cash|imprest|suspense|contra|ledger|narration|fiscal|shrawan|ashadh|ird|ocr|pan\b|esewa|khalti)\b|[\u0900-\u097F]/i;

const FRAMEWORK_TERMS =
  /\b(faithful\s*representation|biswasilo|pratinidhitwo|sambandhitata|relevance|recognition\s*criteria|manyata|mulyankan|nyaya\s*mulya|going\s*concern|chalirakhne|accrual\s*basis|prapti\s*aadhar|conceptual\s*framework|qualitative\s*characteristics|capital\s*maintenance|derecognition|measurement\s*basis|substance\s*over\s*form|materiality|unit\s*of\s*account|present\s*obligation|economic\s*resource|comprehensive\s*income|performance\s*obligation|contract\s*liability|ifrs\s*para|para\s*\d+\.\d+)\b/i;

const COMPLIANCE_TERMS =
  /\b(ird|vat\s*act|income\s*tax|tds\s*rate|ssf\s*rate|gratuity\s*act|labour\s*act|companies\s*act|presumptive\s*tax|advance\s*tax|tax\s*invoice|blacklist|zero\s*rated|exempt\s*supply|reverse\s*charge|excise|customs|withholding|remittance|form\s*13|anna\s*13|pan\s*number|fiscal\s*year|shrawan|ashadh|bs\s*date|nepal\s*tax|statutory|compliance)\b/i;

const ENTRY_SIGNALS =
  /\b(sold|sale|bought|purchase|paid|received|tiryo|tireko|diye|diyo|kineko|kinyo|kinye|kine|kiniyo|kharid|becheko|bikri|kharcha|expense|salary|talab|vat|tds|depreciation|loan|drawings|capital|stock|discount|provision|accrual|contra|deposit|withdraw|refund|return|write\s*off|recover\w*|commission|advance|opening|jama|tirna|bhugtan|firta|firtayo|rupaya|rupiya)\b.*\d|\d.*\b(sold|sale|bought|purchase|paid|received|tiryo|diye|kineko|kinye|kine|kinyo|becheko|kharcha|salary|vat|tds|loan|capital|stock|discount|return|write\s*off|commission|advance|jama|tiryo|diyo|kin)\b/i;

const META_SYSTEM =
  /\b(am\s+i\s+online|online\s*du|online\s*chu|who\s+are\s+you|what\s+are\s+you|what\s+can\s+you|timi\s+ko\s+ho|ke\s+ho\s+timi|your\s+brain|mero\s+brain|ollama|connected|connection)\b/i;

const EMOTIONAL =
  /\b(dukhi|dukh|sad|angry|rish|frustrated|lonely|akela|thak|tired|worried|chinta|hurt|dard|miss|pyar|love|hate|pasand|feel|mood|joke|funny|movie|momo|khana\s*khay|how\s+are\s+you|k\s*cha\s*timro)\b/i;

const EXTERNAL_FACT =
  /\b(weather|mausam|news|population|capital\s*of|prime\s*minister|\bpm\b|president|who\s+is|where\s+is|when\s+did|football|cricket|movie|actor|celebrity)\b/i;

const QUESTION =
  /\b(what|how|why|when|where|who|which|k\s*ho|ke\s*ho|kasari|kina|kun|kati|define|explain|meaning|matlab|arth|farak|difference|compare|classify|entry|journal)\b|\?/i;

export function classifyDomain(text: string): DomainRouteResult {
  const t = text.trim();
  if (!t) {
    return { domain: "emotional_chat", confidence: 1, blockWebSearch: true };
  }

  if (META_SYSTEM.test(t)) {
    return { domain: "meta_system", confidence: 0.95, blockWebSearch: true };
  }

  // Nepal AI question patterns — "noksan k ho" before journal path
  if (isNepaliAccountingQuestion(t)) {
    return { domain: "accounting_qa", confidence: 0.9, blockWebSearch: true };
  }

  // Semantic transaction detection — understands meaning, not just keyword co-occurrence
  if (isSemanticTransaction(t) && !QUESTION.test(t)) {
    return { domain: "journal_entry", confidence: 0.92, blockWebSearch: true };
  }

  const semanticFrame = parseSemanticFrame(t);
  if (
    semanticFrame.action !== "UNKNOWN" &&
    semanticFrame.amount !== null &&
    !semanticFrame.isQuestion &&
    !semanticFrame.isNegated
  ) {
    return { domain: "journal_entry", confidence: 0.88, blockWebSearch: true };
  }

  if (ENTRY_SIGNALS.test(t) && !QUESTION.test(t)) {
    return { domain: "journal_entry", confidence: 0.9, blockWebSearch: true };
  }

  if (FRAMEWORK_TERMS.test(t)) {
    return { domain: "framework_qa", confidence: 0.88, blockWebSearch: true };
  }

  if (COMPLIANCE_TERMS.test(t)) {
    return { domain: "compliance_qa", confidence: 0.85, blockWebSearch: true };
  }

  if (ACCOUNTING_TERMS.test(t)) {
    const isQuestion = QUESTION.test(t);
    if (isQuestion) {
      return { domain: "accounting_qa", confidence: 0.82, blockWebSearch: true };
    }
    if (/\d/.test(t)) {
      return { domain: "journal_entry", confidence: 0.75, blockWebSearch: true };
    }
    return { domain: "accounting_qa", confidence: 0.7, blockWebSearch: true };
  }

  if (EMOTIONAL.test(t) && !ACCOUNTING_TERMS.test(t)) {
    return { domain: "emotional_chat", confidence: 0.7, blockWebSearch: true };
  }

  if (EXTERNAL_FACT.test(t)) {
    return { domain: "external_fact", confidence: 0.75, blockWebSearch: false };
  }

  if (QUESTION.test(t)) {
    return { domain: "external_fact", confidence: 0.5, blockWebSearch: false };
  }

  return { domain: "emotional_chat", confidence: 0.4, blockWebSearch: true };
}

export function isAccountingDomain(text: string): boolean {
  const route = classifyDomain(text);
  return route.blockWebSearch && route.domain !== "meta_system" && route.domain !== "emotional_chat";
}

export function shouldBlockWebSearch(text: string): boolean {
  return classifyDomain(text).blockWebSearch;
}
